source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -10,28 +10,45 @@ pub mod abi {
     include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
 }
 
-/// Declare the init function for the dynamic module. This function is called when the dynamic module is loaded.
-/// The function must return true on success, and false on failure. When it returns false,
-/// the dynamic module will not be loaded.
+/// Declare the init functions for the dynamic module.
 ///
-/// This is useful to perform any process-wide initialization that the dynamic module needs.
+/// The first argument has [`ProgramInitFunction`] type, and it is called when the dynamic module is loaded.
+///
+/// The second argument has [`NewHttpFilterConfigFunction`] type, and it is called when the new HTTP filter configuration is created.
 ///
 /// # Example
 ///
 /// ```
-/// use envoy_proxy_dynamic_modules_rust_sdk::declare_program_init;
+/// use envoy_proxy_dynamic_modules_rust_sdk::*;
 ///
-/// declare_program_init!(my_program_init);
+/// declare_init_functions!(my_program_init, my_new_http_filter_config_fn);
 ///
 /// fn my_program_init() -> bool {
 ///    true
 /// }
+///
+/// fn my_new_http_filter_config_fn(
+///   _envoy_filter_config: EnvoyHttpFilterConfig,
+///   _name: &str,
+///   _config: &str,
+/// ) -> Option<Box<dyn HttpFilterConfig>> {
+///   Some(Box::new(MyHttpFilterConfig {}))
+/// }
+///
+/// struct MyHttpFilterConfig {}
+///
+/// impl HttpFilterConfig for MyHttpFilterConfig {}
 /// ```
 #[macro_export]
-macro_rules! declare_program_init {
-    ($f:ident) => {
+macro_rules! declare_init_functions {
+    ($f:ident,$new_http_filter_config_fn:expr) => {
         #[no_mangle]
         pub extern "C" fn envoy_dynamic_module_on_program_init() -> *const ::std::os::raw::c_char {
+            unsafe {
+                // We can assume that this is only called once at the beginning of the program, so it is safe to set a mutable global variable.
+                envoy_proxy_dynamic_modules_rust_sdk::NEW_HTTP_FILTER_CONFIG_FUNCTION =
+                    $new_http_filter_config_fn
+            };
```

Comment:
I would just use LazyLock here and avoid unsafe unless there is a good reason.

Reply:
https://github.com/envoyproxy/envoy/pull/37070/commits/f59880cb68bfa4ee2c6f77d9eb89eb17ce2187a6 done